### PR TITLE
Added Pelco-P and added UART Class

### PR DESCRIPTION
**Pull request by LuukVerhagen**
I added the Pelco-P protocol (we still use it in our church) to the plugin.

To Generalize things a bit, I added a separate UART class, instead of writing a PelcoPUart Class. In the future, we should let ViscaUart inherent the UART class.

I don't have much c++ experience (I do have experience in other languages though), so I am open to feedback. I tried to follow your coding style as much as possible.

Signed-off-by: Luuk Verhagen <developer@verhagenluuk.nl>

### CODE REVIEW

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:67]: I would recommend something like this:

```
QByteArray result = QByteArray::fromHex("a000") + msg + QByteArray::fromHex("af00");
result[1] = address - 1;
result[7] = checkSum(result);
```

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:77]: Use a QByteArray here

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:188]: Single line condition. Drop the { } around return.

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:198]: ditto

File: ptz-pelco_p.hpp
Comment (glikely) [inline ptz-pelco_p.hpp:18]: As discussed, this doesn't do anything special. Just use a QByteArray directly.

With VISCA I created a new class because I created encoder/decoder helper classes to make the protocol more extensible. At this stage you're not trying to do that, so adding a class just makes the code more complex.

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:94]: Dead code. Nothing calls timeout(). Remove.

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:33]: Dead code. Nothing emits a reset signal. Remove.

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:55]: Dead code, remove

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:49]: I know this is just a debug stub, but the receive hook should check the address of the returned message to see if it is for this camera instance. Like VISCA, PELCO allows multiple cameras on the same serial port.

File: ptz-pelco_p.cpp
Comment (glikely) [inline ptz-pelco_p.cpp:117]: Move { onto line above.

File: ptz-pelco_p.hpp
Comment (glikely) [inline ptz-pelco_p.hpp:55]: Missing newline at end of file

File: ptz_pelco_p.cpp
Comment (glikely) [inline ptz_pelco_p.cpp:4]: This is a dead file. It needs to be removed from the git commit

File: settings.cpp
Comment (glikely) [inline settings.cpp:101]: I think the viscaPort and viscaAddress fields can be reused for PELCO connections too. So, instead of adding additional fields, change the name on the label to "VISCA Serial" or "PELCO-P Serial" depending on what kind of camera is selected. That will remove the need to double this code up.

File: settings.cpp
Comment (glikely) [inline settings.cpp:118]: This is an unrelated whitespace change. It should be removed from the commit

File: settings.cpp
Comment (glikely) [inline settings.cpp:122]: nit: need space between ')' and '{'

File: settings.cpp
Comment (glikely) [inline settings.cpp:161]: Nit: { on same line as 'if' please

File: settings.cpp
Comment (glikely) [inline settings.cpp:218]: Here you can do:
```
ui->label_6->setText("PELCO-P Serial")
```
And then add similar text to the VISCA section. You'll probably also want to change 'label_6' to something meaningful in Qt Designer

**Review by glikely — CHANGES_REQUESTED** [review]
Hi Luuk,

Thanks for this. I'm keen to get PELCO added to the controller. I've made a number of comments on the code. Overall it is in the right direction. I cannot merge it in as-is, but please address the review comments and resubmit the pull request.

For the next submission, please include a proper commit message and signed-off-by line as per the CONTRIBUTING instructions found here. I'm duplicating the Linux kernel contribution process, and I'll only accept patches that are properly signed off.

https://github.com/glikely/obs-ptz/blob/main/CONTRIBUTING.md

**Comment by glikely** [comment]
> I added the Pelco-P protocol (we still use it in our church) to the plugin.
> 
> To Generalize things a bit, I added a separate UART class, instead of writing a PelcoPUart Class. In the future, we should let ViscaUart inherent the UART class.
> 
> I don't have much c++ experience (I do have experience in other languages though), so I am open to feedback. I tried to follow your coding style as much as possible.
> 
> Signed-off-by: Luuk Verhagen [developer@verhagenluuk.nl](mailto:developer@verhagenluuk.nl)

Just saw your signed-off-by line here in the pull request. Thanks! When you respin the branch to address the comments, make sure each commit message includes a description and signed-off-by line. Otherwise both get lost when I merged the branch.

**Comment by LuukVerhagen** [comment]
Thank you for your extensive comments! I applied all your comments to my code. 

Ps. Sorry for the 4 merge commits :)

**Comment by glikely** [comment]
Hey Luuk,

Just saw you updated the branch. I hope to be able to take a look tomorrow. Don't worry about merge conflicts. I'd much rather handle the conflicts myself when I merge the branch so I can see what is conflicting.

However, I do want the fixups to be squashed back into the original commits so that I've got a clean set of changes to look at. `git rebase --interactive` is your friend here. Shout if you don't know how to do this and I can walk you through it.

**Comment by glikely** [comment]
Hi Luuk,

I've fetched your tree and pruned/squashed the commits to get a clean commit history. I needed to do this so I can see the final results instead of the fixups in a separate commit. You can see the result in this branch here:

https://github.com/glikely/obs-ptz/commits/pelcop-draft
https://github.com/glikely/obs-ptz/network

I've made comments on the squashed commit here. Mostly minor stuff at this point.

https://github.com/glikely/obs-ptz/commit/39e078f723b430c41c3dfe4b48cb711b8fe49c81

Since the remaining issues are minor, it would be no problem for me to handle the fixups and commit the result. However it would be better if you are able to clean up your branch with the fixups squashed in. Then I know that I'm merging tested code and it will make it easier for me to pick up future pull requests from you becuase you'll know what is needed.

Let me know if you want to try respinning the branch, or if you want me to commit my reworked version.

**Comment by LuukVerhagen** [comment]
I merged everything into one commit, as you said. I also fixed your remarks

**Comment by glikely** [comment]
Almost there. :-) Thanks for squashing it all into one commit, but you also squashed in the merge commits which means all the stuff I've added to the main branch is showing up in your diff. Reset your branch to what it was before you merged in my mainline and then squash the commits together. `git reflog` will show you the history for finding the pre-merge commit.

**Comment by glikely** [comment]
You may also find `git cherrypick` helpful to create a branch with only the commits you care about:

```
git show-branch --sha1-name origin/main main. # to show the commits on each branch
git reset --hard ade3dfa # to get back to pre-merge state
git cherry-pick <sha1 of your post-merge fixup, if you have a commot for it>
git rebase --interactive 4847dd9. # and follow instructions to squash fixups into main patch
git show-branch origin/main main # to admire your result!
```

**Comment by glikely** [comment]
Another approach, since you've already merged with my current mainline, rebasing onto that branch will probably get rid of the extra diff hunks:

```
git rebase origin/main
git show
```

**Comment by glikely** [comment]
Hi Luuk,

I'm working on getting the v0.5.0 release out. I want to have your PELCO support included, so I've gone ahead and cleaned up commit to remove the extra hunks and rebase it onto the main tree. You can find the commit here:

https://github.com/glikely/obs-ptz/commit/b83b4414e2712dc4c87979999892140b7eeb0f25

**Comment by LuukVerhagen** [comment]
I've been bussy the last few days, so thank you

**Pull request closed without merging**
